presto-main/src/main/java/com/facebook/presto/operator/project/PageProcessor.java

```diff
@@ -0,0 +1,206 @@
+                return new PageProcessorOutput(page.getRetainedSizeInBytes(), Iterators.singletonIterator(new Page(selectedPositions.size())));
+            }
+
+            if (selectedPositions.size() != page.getPositionCount()) {
+                return new PageProcessorOutput(page.getRetainedSizeInBytes(), new PositionsPageProcessorIterator(session, page, selectedPositions));
+            }
+        }
+
+        return new PageProcessorOutput(
+                page.getRetainedSizeInBytes(),
+                new PositionsPageProcessorIterator(session, page, SelectedPositions.positionsRange(0, page.getPositionCount())));
+    }
+
+    private class PositionsPageProcessorIterator
+            extends AbstractIterator<Page>
+    {
+        private final ConnectorSession session;
+        private final Page page;
+
+        private SelectedPositions selectedPositions;
+        private final Block[] previouslyComputedResults;
+
+        public PositionsPageProcessorIterator(ConnectorSession session, Page page, SelectedPositions selectedPositions)
+        {
+            this.session = session;
+            this.page = page;
+            this.selectedPositions = selectedPositions;
+            this.previouslyComputedResults = new Block[projections.size()];
+        }
+
+        @Override
+        protected Page computeNext()
+        {
+            while (true) {
+                if (selectedPositions.isEmpty()) {
+                    return endOfData();
+                }
+
+                int batchSize = Math.min(selectedPositions.size(), projectBatchSize);
+                Optional<Page> result = processBatch(batchSize);
+
+                // if the page buffer filled up, so halve the batch size and retry
+                if (!result.isPresent()) {
+                    verify(batchSize > 1);
+                    projectBatchSize = projectBatchSize / 2;
+                    continue;
+                }
+
+                Page page = result.get();
+
+                // if we produced a large page, halve the batch size for the next call
+                long pageSize = page.getSizeInBytes();
+                if (page.getPositionCount() > 1 && pageSize > MAX_PAGE_SIZE_IN_BYTES) {
+                    projectBatchSize = projectBatchSize / 2;
+                }
+
+                // if we produced a small page, double the batch size for the next call
+                if (pageSize < MIN_PAGE_SIZE_IN_BYTES && projectBatchSize < MAX_BATCH_SIZE) {
+                    projectBatchSize = projectBatchSize * 2;
+                }
+
+                // remove batch from selectedPositions and previouslyComputedResults
+                selectedPositions = selectedPositions.subRange(batchSize, selectedPositions.size());
+                for (int i = 0; i < previouslyComputedResults.length; i++) {
+                    if (previouslyComputedResults[i] != null && previouslyComputedResults[i].getPositionCount() > batchSize) {
```

Comment:
Some more retrospective comments:
@dain I believe `previouslyComputedResults[i].getPositionCount() > batchSize` should actually be:
`previouslyComputedResults[i].getPositionCount() >= batchSize`. Otherwise you won't reuse second half of buffer after `projectBatchSize` halving.